linera-ethereum/src/provider.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use alloy::{
+    primitives::{Address, U256},
+    providers::{Provider, ProviderBuilder, RootProvider},
+    rpc::types::eth::{
+        request::{TransactionInput, TransactionRequest},
+        Filter,
+    },
+    transports::http::reqwest::{header::CONTENT_TYPE, Client},
+};
+use alloy_primitives::Bytes;
+use async_trait::async_trait;
+use url::Url;
+
+use crate::client::{EthereumQueries, JsonRpcClient};
+
+pub type HttpProvider = RootProvider<alloy::transports::http::Http<Client>>;
+
+use crate::{
+    client::get_block_id,
+    common::{event_name_from_expanded, parse_log, EthereumEvent, EthereumServiceError},
+};
+
+/// The Ethereum endpoint and its provider used for accessing the Ethereum node.
+pub struct EthereumClientSimplified {
+    pub url: String,
+}
+
+#[async_trait]
+impl JsonRpcClient for EthereumClientSimplified {
+    type Error = EthereumServiceError;
+    async fn request_inner(&self, payload: Vec<u8>) -> Result<Vec<u8>, Self::Error> {
+        let res = Client::new()
+            .post(self.url.clone())
+            .body(payload)
+            .header(CONTENT_TYPE, "application/json")
+            .send()
+            .await?;
+        let body = res.bytes().await?;
+        Ok(body.as_ref().to_vec())
```

Comment:
If it's always JSON anyway, we could just return `Ok(res.json()?)` here?

And if we made the trait method's argument any `T: Serialize`, we can just use `RequestBuilder::json`.

(Would it even make sense to make the `fetch_json` runtime function less general, too, and turn it into a JSON-RPC oracle? Maybe something to discuss with the team.)

Reply:
I tried that (more precisely with `Ok(res.json().await?)` and no, it does not work.

Reply:
Why? We'd have to change the function's return type, of course.

Reply:
The thing is that there is no JSON in this part of the code. The json serialization/deserialization is happening elsewhere.